addon/mixins/adapter-fetch.js

```diff
@@ -179,14 +179,20 @@ export default Ember.Mixin.create({
 
     return this._ajaxRequest(hash)
       .catch((error, response, requestData) => {
-        throw this.ajaxError(error, response, requestData);
+        throw this.ajaxError(this, response, requestData, error);
       })
       .then((response) => {
+        return RSVP.hash({
+          responseData: determineBodyPromise(response, requestData),
+          response
+        });
+      })
+      .then(({response, responseData}) => {
```

Comment:
I'll add a test for this, but that is the main thing i was wanting to change his n all this. Without waiting for the response of an error to be parsed, there is no way to requests with failing status codes that contain a meaningful response body.